islandora_document.module

```diff
@@ -141,45 +141,88 @@ function islandora_document_islandora_sp_document_islandora_ingest_steps() {
 /**
  * Implements hook_CMODEL_PID_islandora_derivative().
  */
-function islandora_document_islandora_sp_document_islandora_derivative() {
+function islandora_document_islandora_sp_document_islandora_derivative($object) {
   $derivatives = array();
-  $derivatives[] = array(
-    'source_dsid' => 'OBJ',
-    'destination_dsid' => 'PDF',
-    'weight' => '0',
-    'function' => array(
-      'islandora_document_add_pdf_derivative',
-    ),
-    'file' => drupal_get_path('module', 'islandora_document') . '/includes/derivatives.inc',
-  );
-  $derivatives[] = array(
-    'source_dsid' => 'PDF',
-    'destination_dsid' => 'TN',
-    'weight' => '0',
-    'function' => array(
-      'islandora_document_add_tn_derivative',
-    ),
-    'file' => drupal_get_path('module', 'islandora_document') . '/includes/derivatives.inc',
-  );
-  $derivatives[] = array(
-    'source_dsid' => 'PDF',
-    'destination_dsid' => 'PREVIEW',
-    'weight' => '1',
-    'function' => array(
-      'islandora_document_add_preview_derivative',
-    ),
-    'file' => drupal_get_path('module', 'islandora_document') . '/includes/derivatives.inc',
-  );
-  if (variable_get('islandora_document_create_fulltext', FALSE)) {
+  if ($object['OBJ']) {
+    $derivatives[] = array(
+      'source_dsid' => 'OBJ',
+      'destination_dsid' => 'PDF',
+      'weight' => '0',
+      'function' => array(
+        'islandora_document_add_pdf_derivative',
+      ),
+      'file' => drupal_get_path('module', 'islandora_document') . '/includes/derivatives.inc',
+    );
+    $derivatives[] = array(
+      'source_dsid' => 'PDF',
+      'destination_dsid' => 'TN',
+      'weight' => '0',
+      'function' => array(
+        'islandora_document_add_tn_derivative',
+      ),
+      'file' => drupal_get_path('module', 'islandora_document') . '/includes/derivatives.inc',
+    );
     $derivatives[] = array(
       'source_dsid' => 'PDF',
-      'destination_dsid' => 'FULL_TEXT',
-      'weight' => '2',
+      'destination_dsid' => 'PREVIEW',
+      'weight' => '1',
       'function' => array(
-        'islandora_document_add_fulltext_derivative',
+        'islandora_document_add_preview_derivative',
       ),
       'file' => drupal_get_path('module', 'islandora_document') . '/includes/derivatives.inc',
     );
+    if (variable_get('islandora_document_create_fulltext', FALSE)) {
+      $derivatives[] = array(
+        'source_dsid' => 'PDF',
+        'destination_dsid' => 'FULL_TEXT',
+        'weight' => '2',
+        'function' => array(
+          'islandora_document_add_fulltext_derivative',
+        ),
+        'file' => drupal_get_path('module', 'islandora_document') . '/includes/derivatives.inc',
+      );
+    }
   }
+  else {
```

Comment:
Seems like this could be solved by throwing the `optional="true"` flag into the DS-COMP (even though it may not strictly be true)... Should [make the `PDF` stream get skipped in the importer](https://github.com/Islandora/islandora_importer/blob/7.x/modules/zip_importer/includes/importer.inc#L283-L290)...

I'm not particularly comfortable with this implementation using the derivative hook, since we don't use the `source_dsid` value when running the given function (and the logic would may have to change quite a bit in order to make it happen)... We'd see we have an `OBJ` stream, and so trigger a derivative function where we try to grab the `PDF`. It may be that we'd never see this in practice, but... still. It feels unsafe.

... The "real" solution would be reworking how we map files to datastreams, to make things consistent... Blargh.